parity-clib/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Note that all the structs and functions here are documented in `parity.h`, to avoid
+//! duplicating documentation.
+
+extern crate parity;
+
+use std::os::raw::{c_char, c_void, c_int};
+use std::panic;
+use std::ptr;
+use std::slice;
+
+#[repr(C)]
+pub struct ParityParams {
+	pub configuration: *mut c_void,
+	pub on_client_restart_cb: Option<extern "C" fn(*mut c_void, *const c_char, usize)>,
+	pub on_client_restart_cb_custom: *mut c_void,
+}
+
+#[no_mangle]
+pub extern fn parity_config_from_cli(args: *const *const c_char, args_lens: *const usize, len: usize, output: *mut *mut c_void) -> c_int {
```

Comment:
isn't `pub unsafe extern "C" fn ...` signature recommended for c api?

Reply:
According to the official docs, no: https://doc.rust-lang.org/nomicon/ffi.html#calling-rust-code-from-c
I agree that this is really debatable, though.

Reply:
Right, i remember i was given this recommendation by Rust guys though
nvm